sdk/src/azure/core/az_mqtt5_rpc.c

```diff
@@ -21,6 +23,249 @@ static const az_span command_name_key = AZ_SPAN_LITERAL_FROM_STR(AZ_MQTT5_RPC_CO
+      || az_span_is_content_equal(value, extracted_token))
+  {
+    if (out_token)
+    {
+      *out_token = extracted_token;
+    }
+    return AZ_OK;
+  }
+  return AZ_ERROR_IOT_TOPIC_NO_MATCH;
+}
+
+AZ_NODISCARD az_result _az_mqtt5_rpc_replace_tokens_in_format(
+    az_span mqtt_topic_span,
+    az_span topic_format,
+    az_span service_group_id,
+    az_span client_id,
+    az_span service_id,
+    az_span executor_id,
+    az_span command_id,
+    uint32_t* required_length)
+{
+  az_result ret = AZ_OK;
+  *required_length += (int32_t)sizeof((uint8_t)'\0');
+  uint32_t buffer_length = (uint32_t)az_span_size(mqtt_topic_span);
+  az_span remainder = mqtt_topic_span;
+
+  if (!az_span_is_content_equal(service_group_id, AZ_SPAN_EMPTY))
+  {
+    *required_length += (uint32_t)az_span_size(service_group_id_key)
+        + (uint32_t)az_span_size(service_group_id) + 1;
+    _az_RETURN_IF_NOT_ENOUGH_SIZE(mqtt_topic_span, (int32_t)*required_length);
+    remainder = az_span_copy(remainder, service_group_id_key);
+    remainder = az_span_copy(remainder, service_group_id);
+    remainder = az_span_copy_u8(remainder, '/');
+  }
+
+  for (int32_t i = 0; i < az_span_size(topic_format); i++)
+  {
+    uint8_t c = az_span_ptr(topic_format)[i];
+    _az_RETURN_IF_NOT_ENOUGH_SIZE(mqtt_topic_span, (int32_t)(++*required_length));
+
+    if (c == '{')
+    {
+      if (i + 1 >= az_span_size(topic_format))
+      {
+        return AZ_ERROR_ARG;
+      }
+      az_span token
+          = _az_mqtt5_rpc_extract_next_token_from_format(az_span_slice_to_end(topic_format, i + 1));
+      if (az_span_is_content_equal(token, AZ_SPAN_EMPTY))
+      {
+        return AZ_ERROR_ARG;
+      }
+      uint32_t curr_hash = _az_mqtt5_rpc_calculate_hash(token);
+      if (curr_hash == az_mqtt5_rpc_client_id_hash)
+      {
+        ret = _az_mqtt5_rpc_verify_buffer_length_and_copy(
+            client_id, &remainder, required_length, buffer_length);
+        i += az_span_size(invoker_client_id_key) - 1;
+      }
+      else if (curr_hash == az_mqtt5_rpc_service_id_hash)
+      {
+        ret = _az_mqtt5_rpc_verify_buffer_length_and_copy(
+            service_id, &remainder, required_length, buffer_length);
+        i += az_span_size(model_id_key) - 1;
+      }
+      else if (curr_hash == az_mqtt5_rpc_executor_id_hash)
+      {
+        ret = _az_mqtt5_rpc_verify_buffer_length_and_copy(
+            executor_id, &remainder, required_length, buffer_length);
+        i += az_span_size(executor_client_id_key) - 1;
+      }
+      else if (curr_hash == az_mqtt5_rpc_command_id_hash)
+      {
+        ret = _az_mqtt5_rpc_verify_buffer_length_and_copy(
+            command_id, &remainder, required_length, buffer_length);
+        i += az_span_size(command_name_key) - 1;
+      }
+      else
+      {
+        return AZ_ERROR_ARG;
+      }
+    }
+    else
+    {
+      remainder = az_span_copy_u8(remainder, c);
+    }
+  }
+
+  az_span_copy_u8(remainder, '\0');
```

Comment:
I'm not sure the `required_length` is accurate. I think it may be missing the remainder and the NULL terminator.

The usage pattern for the codec is that:
- If the buffer is large enough, then we'll fill it in and return AZ_OK. `required_length` will contain how much we've filled in.
- If the buffer is too small, then we
  1. Guarantee that we do not overrun the buffer
  2. Guarantee that we correctly calculate the `required_length`

The usage pattern for an _allocating_ application is to start with the existing buffer: if the call fails then the application can reallocate a buffer of `required_length` and try again.

Reply:
I have added some tests to ensure that `required_length` is accurate, I add the null terminator size at the start of the function.